core/templates/pages/exploration-editor-page/improvements-tab/improvements-tab.component.ts

```diff
@@ -13,10 +13,117 @@
+        ExplorationImprovementsTaskRegistryService, RouterService,
+        UrlInterpolationService) {
+      this.$onInit = () => {
+        this.isInitialized = false;
+        this.timeMachineImageUrl = (
+          UrlInterpolationService.getStaticImageUrl('/icons/time_machine.svg'));
+        this.navigateToStateEditor = (stateName: string) => {
+          RouterService.navigateToMainTab(stateName);
+        };
+
+        // AngularJS immediately refreshes directives after $q promises fulfill,
+        // so we convert to one from the "plain" promise returned by the
+        // service.
+        $q.when(ExplorationImprovementsService.isImprovementsTabEnabledAsync())
+          .then(improvementsTabIsEnabled => {
+            if (!improvementsTabIsEnabled) {
+              return;
+            }
+
+            const expStats = (
+              ExplorationImprovementsTaskRegistryService.getExplorationStats());
+            this.completionRate = (
+              (expStats && expStats.numStarts > 0) ?
+                (expStats.numCompletions / expStats.numStarts) : 0);
+            this.completionRateAsPercent = (
+              Math.round(100.0 * this.completionRate) + '%');
+
+            const hbrTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenHighBounceRateTasks());
+            const iflTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenIneffectiveFeedbackLoopTasks());
+            const ngrTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenNeedsGuidingResponsesTasks());
+            const siaTasks = (
+              ExplorationImprovementsTaskRegistryService
+                .getOpenSuccessiveIncorrectAnswersTasks());
+            this.getNumTasks = () => (
+              hbrTasks.length + iflTasks.length + ngrTasks.length +
+              siaTasks.length);
+            this.getNumExpLevelTasks = () => hbrTasks.length;
+            this.getNumConceptLevelTasks = () => iflTasks.length;
+            this.getNumCardLevelTasks = () => ngrTasks.length + siaTasks.length;
+            this.hasCriticalTasks = () => ngrTasks.length > 0;
+            this.getExplorationHealth = () => {
+              if (this.hasCriticalTasks()) {
+                return ImprovementsConstants.EXPLORATION_HEALTH_TYPE_CRITICAL;
+              } else if (this.getNumTasks() > 0) {
+                return ImprovementsConstants.EXPLORATION_HEALTH_TYPE_WARNING;
+              } else {
+                return ImprovementsConstants.EXPLORATION_HEALTH_TYPE_HEALTHY;
+              }
+            };
+
+            this.stateTasksObject = (
+              ExplorationImprovementsTaskRegistryService
+                .makeStateTasksObject());
+
+            const stateRetentions: Map<string, number> = new Map(
+              Object.keys(this.stateTasksObject).map(stateName => {
+                const {numCompletions, totalHitCount} = (
```

Comment:
Can you talk through this notation a bit? expStats.getStateStats() results in a StateStats domain object, right? So I think you should make a variable for that domain object, and then get its fields separately...

Reply:
Oh, I see what's going on, thanks. Yes, I think I would prefer dropping the destructuring, so that you have:

```
const stateStats = expStats.getStateStats(stateName);
const numCompletions = stateStats.numCompletions;
const totalHitCount = stateStats.totalHitCount;
```

I realize it's a line longer, but IMO it is easier to follow.